### PR TITLE
Set the availability of the VideoPress extension for Gutenberg

### PR DESCRIPTION
**Pull request by mmtr**
### Changes proposed in this Pull Request:

This PR sets the `jetpack/videopress` extension availability, so we can check it on the editor when deciding if we need to enhance the core video block with VideoPress support ([#](https://github.com/Automattic/wp-calypso/pull/30546/files#diff-e436b8d5ae66471d9bff239c6c650f68R61)). This will also allow us in the future to display upgrade nudges if we detect we're missing a plan.

The extension is set available:
* On Jetpack sites having the VideoPress module activated (included in Premium and above subscriptions).
* On WordPress.com sites having the video upgrade (included in Premium and above plans).

### Testing instructions:

**Jetpack**

* Apply this branch to your local environment or [create a new JN site using it](https://jurassic.ninja/create?jetpack-beta&branch=update/set-videopress-extension-availability&shortlived&wp-debug-log&gutenpack).
* Start with a free subscription.
* Go to Settings → Jetpack Constants in the site’s wp-admin and set `JETPACK_BETA_BLOCKS` to true.
* Go to Posts > Add New.
* Open the browser console and type `Jetpack_Editor_Initial_State.available_blocks`.
* Check that the `videopress` extension is marked as not available because we're missing a plan:

screen shot 2019-02-11 at 12 31 46

* Upgrade now to a premium subscription.
* Go to Posts > Add New and check again the `Jetpack_Editor_Initial_State.available_blocks` output.
* Make sure the extension is unavailable yet with `missing_module` as  `unavailable_reason`.
* Go to Jetpack > Settings and activate the VideoPress module (by enabling the "Enable high-speed, ad-free video player" option).
* Go to Posts > Add New and check again the `Jetpack_Editor_Initial_State.available_blocks` output.
* Confirm the extension is available now.
* Go to Settings → Jetpack Constants in the site’s wp-admin and set `JETPACK_BETA_BLOCKS` to false.
* Make sure the extension is not available because it is not whitelisted.

**WordPress.com**

* Apply the diff D24180-code to your sandbox site.
* Start with a free plan.
* Go to Posts > Add New.
* Open the browser console and type `Jetpack_Editor_Initial_State.available_blocks`.
* Check that the `videopress` extension is marked as not available because we're missing a plan.
* Upgrade now to a premium subscription.
* Go to Posts > Add New and check again the `Jetpack_Editor_Initial_State.available_blocks` output.
* Confirm the extension is available now.

#### Proposed changelog entry for your changes:

* VideoPress: Set the Block editor extension availability

### CODE REVIEW

File: modules/videopress/class.videopress-gutenberg.php [hunk 1]
```diff
@@ -46,17 +86,21 @@ public static function register_video_block_with_videopress() {
 	 *
 	 * @return string
 	 */
-	public static function render_video_block_with_videopress( $attributes, $content ) {
+	public function render_video_block_with_videopress( $attributes, $content ) {
```
Comment (ockham) [inline modules/videopress/class.videopress-gutenberg.php]: Why the change to a non-static method? There still aren't any references to `$this` in this methode, are there?

File: modules/videopress/class.videopress-gutenberg.php [hunk 2]
```diff
@@ -11,29 +11,69 @@
...
+
+		// It is available on Jetpack Sites having the VideoPress module active.
+		if ( method_exists( 'Jetpack', 'is_active' ) && Jetpack::is_active() ) {
+			if ( Jetpack::is_module_active( 'videopress' ) ) {
+				Jetpack_Gutenberg::set_extension_available( 'jetpack/videopress' );
+			} elseif ( ! Jetpack::active_plan_supports( 'videopress' ) ) {
+				Jetpack_Gutenberg::set_extension_unavailable( 'jetpack/videopress', 'missing_plan' );
+			} else {
+				Jetpack_Gutenberg::set_extension_unavailable( 'jetpack/videopress', 'missing_module' );
+			}
```
Comment (ockham) [inline modules/videopress/class.videopress-gutenberg.php]: For comparison, this is the check that we use in Simple Payments to determine whether to set the block as available:

https://github.com/Automattic/jetpack/blob/f836eb1f648cb7a8a67573cd208c796c0429a5c0/modules/simple-payments/simple-payments.php#L91-L116

It would be nice to have a helper function for these recurring checks -- ideally by streamlining the WP.com and JP checks for the presence of a given feature. This is probably beyond the scope of this PR -- but for now, do you think it'd make sense to re-use the implementation of that check in Simple Payments (essentially by copy-pasting) to be able to locate these checks more easily later on?

File: modules/videopress/class.videopress-gutenberg.php (hunk 1 above)
Comment (mmtr) [reply modules/videopress/class.videopress-gutenberg.php]: That's true! It's unrelated to what the PR tries to solve, but decided to avoid `static` methods based on the feedback received on p7H4VZ-2bw-p2 (#comment-7379):

> Also avoiding `static` methods as much as possible if they do anything more than calculate and return a value in themselves as they can’t be mocked too.

This is actually don't doing anything more than that, but I think it's worth the change in order to have consistency with the other methods and remove any potential issue we may have if we want to add unit tests.

File: modules/videopress/class.videopress-gutenberg.php (hunk 2 above)
Comment (mmtr) [reply modules/videopress/class.videopress-gutenberg.php]: When you say re-use the implementation, do you mean to have a `is_enabled_jetpack_videopress` equivalent function that uses `has_any_blog_stickers` for WP.com and `active_plan_supports` for Jetpack?

Actually, my first attempt was to encapsulate the check in a separate `is_enabled_jetpack_videopress` method so we could re-use it later.

But then I realized this is not suitable for VideoPress because `active_plan_supports` returns true if we have the right plan but the module is inactive. Since the `class.videopress-gutenberg.php` file is loaded always, even if the module is not active (https://github.com/Automattic/jetpack/pull/11268), we cannot rely on the default unavailability reason set by `Jetpack_Gutenber::get_availability`:
 https://github.com/Automattic/jetpack/blob/f836eb1f648cb7a8a67573cd208c796c0429a5c0/class.jetpack-gutenberg.php#L357

Also, I remembered the @dmsnell's reminder that good APIs are discovered, not designed (p7H4VZ-29X-p2), so why don't we develop the helper function in the moment we need to re-use the same in a different place?

File: modules/videopress/class.videopress-gutenberg.php (hunk 1 above)
Comment (ockham) [reply modules/videopress/class.videopress-gutenberg.php]: Hmm, I see. I think I have yet to wrap my head around the implications for mocking (I'll probably ask on that p2 thread), but generally, if a method doesn't make use of `$this`, I think it makes sense to reflect that by making it `static`. This is my way of understanding minimalism: We don't claim to need what we don't use.

This discussion might expose wider architectural considerations -- e.g. singleton patterns (which inevitably imply `static` `init()` methods) are quite widespread in the Jetpack codebase and are somewhat hard to isolate, e.g. for unit tests -- compared e.g. to 'proper' use of OOP patterns (where constructors set up invariants that are cleaned up after by destructors) -- see our recent discussion (https://github.com/Automattic/jetpack/pull/11212#issuecomment-461770189).

I'm not feeling very strongly about this, so if you'd like to go with non-static, feel free to keep the changes.

File: modules/videopress/class.videopress-gutenberg.php (hunk 2 above)
Comment (ockham) [reply modules/videopress/class.videopress-gutenberg.php]: > When you say re-use the implementation, do you mean to have a `is_enabled_jetpack_videopress` equivalent function that uses `has_any_blog_stickers` for WP.com and `active_plan_supports` for Jetpack?

Yeah, something like that for starters -- which we could then extract, parametrize, and re-use in Simple Payments.

> Actually, my first attempt was to encapsulate the check in a separate `is_enabled_jetpack_videopress` method so we could re-use it later.

:heart: 

> But then I realized this is not suitable for VideoPress because `active_plan_supports` returns true if we have the right plan but the module is inactive.

In cases like this, I think it makes sense to ask if these are principle differences, or more of an implementation detail. I like to approach this top-down/semantically: Ideally, we want a helper that can tell us across JP and WP.com if a feature that is part of a paid plan. Modules are similar enough to each other that their activation state shouldn't preclude us from writing such a helper (if anything, the entire concept of a module is something that might not hold up as well in a Gutenberg world, but that's for a different conversation).

So specifically, will it change the block availability behavior of Simple Payments if we add the same module availability check to `is_enabled_jetpack_videopress` that you're using here for VideoPress? (This isn't something you'd need to do as part of this PR -- it's more about starting to converge existing, specific, helper functions, into a shared, generic one.)

> Since the `class.videopress-gutenberg.php` file is loaded always, even if the module is not active (#11268), we cannot rely on the default unavailability reason set by `Jetpack_Gutenber::get_availability`:
> [jetpack/class.jetpack-gutenberg.php](https://github.com/Automattic/jetpack/blob/f836eb1f648cb7a8a67573cd208c796c0429a5c0/class.jetpack-gutenberg.php#L357)
> 
> Line 357 in [f836eb1](/Automattic/jetpack/commit/f836eb1f648cb7a8a67573cd208c796c0429a5c0)
> 
>  				$reason = isset( self::$availability[ $extension ] ) ? self::$availability[ $extension ] : 'missing_module'; 
> Also, I remembered the @dmsnell's reminder that good APIs are discovered, not designed ([p7H4VZ-29X-p2](https://href.li/?https://wp.me/p7H4VZ-29X)), so why don't we develop the helper function in the moment we need to re-use the same in a different place?

I agree with Dennis here, but I'm not sure I'm reading that the way you meant it -- IMO that moment has come, now that both Simple Payments and VideoPress need the same kind of feature/plan check :grimacing:

File: modules/videopress/class.videopress-gutenberg.php [hunk 3]
```diff
@@ -46,17 +86,21 @@ public static function register_video_block_with_videopress() {
...
 			$blog_id = get_current_blog_id();
-		} else {
+		} elseif ( method_exists( 'Jetpack', 'is_active' ) && Jetpack::is_active() ) {
 			$blog_id = Jetpack_Options::get_option( 'id' );
 		}
 
+		if ( ! isset( $blog_id ) ) {
+			return $content;
+		}
+
```
Comment (ockham) [inline modules/videopress/class.videopress-gutenberg.php]: I'm a bit confused about L94-L103; shouldn't `get_current_blog_id()` work for both WP.com and JP alike?

(Picking the right blog ID seems like such a universal problem that I'd hope we have code that addresses it generically -- otherwise, pretty much every file in JP would have to include logic like the above, wouldn't it? Which I'd find somewhat hard to believe...)

**Review by ockham — DISMISSED** [review]
Left a few questions, some of which are of a more abstract/architectural nature and don't need to be addressed as part of this PR, so don't let them block you. (I'd only be curious to hear your thoughts on https://github.com/Automattic/jetpack/pull/11321/files#r255582408)

Testing-wise, this looks good. I've noticed that if I switch back to a Free plan, the VideoPress block remains available -- probably because of #11268, so I guess that's deliberate(?)

File: modules/videopress/class.videopress-gutenberg.php (hunk 2 above)
Comment (mmtr) [reply modules/videopress/class.videopress-gutenberg.php]: Wonderful, thanks @ockham! I have a clearer perspective now :)

> In cases like this, I think it makes sense to ask if these are principle differences, or more of an implementation detail. I like to approach this top-down/semantically: Ideally, we want a helper that can tell us across JP and WP.com if a feature that is part of a paid plan. Modules are similar enough to each other that their activation state shouldn't preclude us from writing such a helper.

That totally makes sense and indeed the module check won't change the behavior of Simple Payments.

Let me give it a thought and see if I can change this function so we can start reusing it on future iterations.

> IMO that moment has come, now that both Simple Payments and VideoPress need the same kind of feature/plan check

I see it now :) I was considering them different checks!

File: modules/videopress/class.videopress-gutenberg.php (hunk 3 above)
Comment (mmtr) [reply modules/videopress/class.videopress-gutenberg.php]: I always got the correct id when using `get_current_blog_id` but @dmsnell was getting unexpected values which were solved when using `Jetpack_Options::get_option( 'id' )`  ([#](https://github.com/Automattic/jetpack/pull/11193#issuecomment-457883886), [#](https://github.com/Automattic/jetpack/pull/11193/commits/215cf789f3d8bd03ff9eb1bbdb693acb8831d273)).

Comment (ockham) [reply modules/videopress/class.videopress-gutenberg.php]: Hmm, I've seen some precedent where -- in code invoked through a REST API endpoint -- `get_current_blog_id()` would return `public-api.wordpress.com`'s site ID. This happened most likely since it was called through a hook that was triggered before the relevant `switch_to_blog()` call, and could be fixed by invoking the relevant action from the function that was called to provide the endpoint's response.

I don't think I've personally seen `get_current_blog_id()` return `1` yet :thinking: I wonder if that could be prevented by setting a later priority to the `add_action` call that attaches the enclosing function to `init`. Not that guessing priorities is exactly elegant, but it's arguably kinda WordPress-y :stuck_out_tongue: 

Anyway, if you keep it the way it is now, would you mind adding an inline comment to the code explaining why we're not using `get_current_blog_id()` in the Jetpack case? (i.e. essentially what you explained in https://github.com/Automattic/jetpack/pull/11321#discussion_r255600102.)

**Comment by mmtr** [comment]
Thanks for the review @ockham!

> Testing-wise, this looks good. I've noticed that if I switch back to a Free plan, the VideoPress block remains available -- probably because of #11268, so I guess that's deliberate(?)

That's right. We keep working video blocks enhanced with VideoPress that are included in already published posts, even if you downgrade your plan.

File: modules/videopress/class.videopress-gutenberg.php (hunk 2 above)
Comment (mmtr) [reply modules/videopress/class.videopress-gutenberg.php]: In b7659fc, I encapsulated the check logic in a separated `check_videopress_availability` method. 

I have not created this as a global helper because it needs to be also available in WordPress.com but I'm not able to find any file synced with Fusion that is a good candidate to place this function (all the synced files are hard related to specific modules).

We should create a new file where we can include these helpers functions and add it to Fusion, but since this PR has already a counterpart a diff created by Fusion, I preferred to leave that for a follow-up PR.

File: modules/videopress/class.videopress-gutenberg.php (hunk 3 above)
Comment (mmtr) [reply modules/videopress/class.videopress-gutenberg.php]: Done in 117d95b

File: modules/videopress/class.videopress-gutenberg.php (hunk 2 above)
Comment (mmtr) [reply modules/videopress/class.videopress-gutenberg.php]: Actually, I did some tests and I'm afraid that there are some differences between VideoPress and Simple Payments that won't allow us to re-use the helper with the current implementation.

In Jetpack, when I add a Premium Subscription, Jetpack::is_module_active( 'simple-payments' ) returns `false`. But the module is actually enabled and I can add Simple Payments buttons. This means that right now we cannot rely on `is_module_active` for Simple Payments but we need it for VideoPress.

@rodrigoi do you know why the Simple Payments module doesn't get activated in Jetpack?

**Comment by jeherve** [comment]
It does not seem to work for me right now (tested on a JN site). The block remains unavailable, at all times:

```
Jetpack_Editor_Initial_State.available_blocks.videopress
{available: false, unavailable_reason: "not_whitelisted"}
```

Purchasing a plan or enabling the module does not change that.

File: modules/videopress/class.videopress-gutenberg.php (hunk 2 above)
Comment (jeherve) [reply modules/videopress/class.videopress-gutenberg.php]: > do you know why the Simple Payments module doesn't get activated in Jetpack?

`simple-payments` is not a module I am afraid. It's one of the "theme tools" that are loaded regardless of module status, here:
https://github.com/Automattic/jetpack/blob/661d73f2f4da09d8520274fa3b641b97fc30db90/modules/module-extras.php#L25

**Comment by mmtr** [comment]
> It does not seem to work for me right now (tested on a JN site). The block remains unavailable, at all times:
> 
> ```
> Jetpack_Editor_Initial_State.available_blocks.videopress
> {available: false, unavailable_reason: "not_whitelisted"}
> ```
> Purchasing a plan or enabling the module does not change that.

Maybe you don't have enabled the beta blocks. Can you confirm? My bad, I will update the testing instructions on the PR description.

**Comment by jeherve** [comment]
Oh, my bad, I should have thought about that. That was it! I'll carry on with testing now.

File: modules/videopress/class.videopress-gutenberg.php (hunk 2 above)
Comment (mmtr) [reply modules/videopress/class.videopress-gutenberg.php]: Thanks @jeherve.

In that case we need to perform different checks and then I'm not sure if we should go ahead with the re-usable helper. 

We could work-around these differences by indicating via params if we should check the module or not:

```php
function jetpack_check_module_availability( $module, $check_module_is_active = true ) {
	if ( method_exists( 'Jetpack', 'is_active' ) && Jetpack::is_active() ) {
		if ( $check_module_is_active && Jetpack::is_module_active( $module ) ) {
			return array( 'available' => true );
		} elseif ( ! $check_module_is_active && Jetpack::active_plan_supports( $module )  ) {
			return array( 'available' => true );
		} elseif ( ! Jetpack::active_plan_supports( $module ) ) {
			return array(
				'available'          => false,
				'unavailable_reason' => 'missing_plan',
			);
		} else {
			return array(
				'available'          => false,
				'unavailable_reason' => 'missing_module',
			);
		}
	}

	//...
}
```

But these are the kind of abstractions that can lead to a code more difficult to debug.

**Review by jeherve — DISMISSED** [review]
This seems to test well! It should be good to merge now.

I could not test the WordPress.com end though, as the patch does not apply cleanly anymore. That is probably because this PR needs a rebase, because #11310 has landed in Jetpack and on WordPress.com.

File: modules/videopress/class.videopress-gutenberg.php (hunk 2 above)
Comment (jeherve) [reply modules/videopress/class.videopress-gutenberg.php]: At this point I wonder if we should reconsider our approach, and make Simple Payments a module that we auto-activate when you purchase a plan. It may make things easier to handle.

**Review by jeherve — DISMISSED** [review]
Another approval post-rebase.

**Review by kwight — DISMISSED** [review]
I ran through this with a Professional plan, everything worked as expected. Wonderful job! <3

**Comment by kwight** [comment]
> Changelog: * VideoPress: Set the Gutenberg extension availability

@jeherve Small question, are we using the term "Gutenberg" when referring to the editor publicly? Should this be "block editor"?

**Comment by jeherve** [comment]
We usually use "Block Editor".

**Comment by mmtr** [comment]
> We usually use "Block Editor".

Cool. Will update it in the next commit trying to solve the failing unit tests in WordPress.com

**Review by jeherve — DISMISSED** [review]

**Review by jeherve — APPROVED** [review]
This should be good to go! 🚢
